src/uavariant.cpp

```diff
@@ -488,24 +488,112 @@ UaString UaVariant::toFullString() const
-        return OpcUa_Bad;  // Incompatible data type
+		LOG(Log::DBG) << __FUNCTION__ << " conversion failed, function only handles numeric conversions. Cannot convert internal variant type ["<<m_impl->type->typeName<<"]";
+		return OpcUa_Bad;
     }
 
-    *out = *static_cast<T*>(m_impl->data);
-    return OpcUa_Good;
+    return convertNumericType(out);
+}
+
+template<typename TTargetNumericType>
+  UaStatus UaVariant::convertNumericType(TTargetNumericType* out ) const
+{
+	try
+	{
+		switch(m_impl->type->typeIndex)
+		{
+			case UA_TYPES_BOOLEAN:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_Boolean*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_BYTE:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_Byte*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_SBYTE:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_SByte*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_INT16:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_Int16*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_UINT16:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_UInt16*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_INT32:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_Int32*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_UINT32:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_UInt32*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_INT64:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_Int64*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_UINT64:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_UInt64*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_FLOAT:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_Float*>(m_impl->data)) );
+				return OpcUa_Good;
+			case UA_TYPES_DOUBLE:
+				*out = boost::numeric_cast<TTargetNumericType>( *(static_cast<OpcUa_Double*>(m_impl->data)) );
+				return OpcUa_Good;
+			default:
+				LOG(Log::DBG) << __FUNCTION__ << " conversion failed, no conversion handler for internal variant type ["<<m_impl->type->typeName<<"]";
+				return OpcUa_Bad;
+		}
+	}
+	catch(const boost::numeric::bad_numeric_cast& e)
+	{
+		LOG(Log::DBG) << __FUNCTION__ << " conversion failed for internal variant type ["<<m_impl->type->typeName<<"], boost conversion threw exception: " << e.what();
+		return OpcUa_BadOutOfRange;
+	}
+	catch(const std::runtime_error& e)
+	{
+		LOG(Log::DBG) << __FUNCTION__ << " conversion failed for internal variant type ["<<m_impl->type->typeName<<"], conversion threw exception: " << e.what();
+		return OpcUa_Bad;
+	}
+}
+
+bool UaVariant::isNumericType( const UA_DataType& dataType ) const
+{
+	switch(dataType.typeIndex)
+	{
+	case UA_TYPES_BOOLEAN:
```

Comment:
Why boolean is considered a numeric type?

Reply:
i'd say it is no? has value 1 or 0. I'll add a unit test to make sure boost::numeric_cast shares my opinion (i.e. can convert booleans to other numeric types and vice-versa).

Reply:
It's OK for me to convert a boolean to numeric  (0/1), fine - this is not ambiguous.
Where I feel less comfortable:
- that vice-versa direction is applied and that is not so fun anymore (why we have a separate bool type in C++ rather than plain in like in C days?) Plus all the ambiguity that you no longer have 0 and 1.
- that the function called "isNumeric" would be used in other applications than your current one with unpredictable results
- if boolean was numeric, it would have all the numeric operators, but it doesn't. So it's not numeric really ;-)